src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,360 @@
+by simp [←ite_and]
+
+end to_move
+
+namespace simple_graph
+
+variables (G : simple_graph V)
+
+/-! ### Helping lemmas for edges -/
+
+lemma incidence_equiv {i : V} {e : G.edge_set} : ↑e ∈ G.incidence_set i ↔ i ∈ (↑e : sym2 V) :=
+by simp only [incidence_set, true_and, set.mem_sep_eq, subtype.coe_prop]
+
+lemma mem_incidence_set_iff {u v w : V} :
+  ⟦(u, v)⟧ ∈ G.incidence_set w ↔ ⟦(u, v)⟧ ∈ G.edge_set ∧ (w = u ∨ w = v) :=
+by { convert iff.rfl, simp only [mem_iff] }
+
+lemma edge_in_two_incidence_sets {i j : V} {e : sym2 V} (H_ne : i ≠ j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j → e = ⟦(i, j)⟧ :=
+begin
+  refine quotient.rec_on_subsingleton e (λ p, _),
+  rcases p with ⟨v, w⟩,
+  rintro ⟨⟨-, H_i⟩, ⟨-, H_j⟩⟩,
+  rw ←elems_iff_eq H_ne,
+  tauto
+end
+
+lemma mem_incidence_sets_iff_eq {i j : V} {e : sym2 V} (h : G.adj i j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j ↔ e = ⟦(i, j)⟧ :=
+begin
+ split,
+  { apply edge_in_two_incidence_sets _,
+  rintro rfl,
+  exact G.loopless i h },
+  { rintro rfl,
+  nth_rewrite 1 eq_swap,
+  simp only [h, G.sym h, mem_incidence_set, and_self] }
+end
+
+lemma adj_iff_exists_edge_coe {i j : V} : G.adj i j ↔ ∃ (e : G.edge_set), ↑e = ⟦(i, j)⟧ :=
+by simp only [mem_edge_set, exists_prop, set_coe.exists, exists_eq_right, subtype.coe_mk]
+
+/-! ## Incidence matrix M -/
+
+section incidence
+
+variables [fintype V] [decidable_eq V] [decidable_rel G.adj] (R) [ring R]
+
+/-- `inc_matrix G R` is the matrix `M` such that `M i e = 1` if vertex `i` is an
+endpoint of the edge `e` in the simple graph `G`, otherwise `M i j = 0`. -/
+def inc_matrix : matrix V (sym2 V) R
+| i e := if e ∈ G.incidence_set i then 1 else 0
+
+@[simp]
+lemma inc_matrix_apply {i : V} {e : sym2 V} :
+  G.inc_matrix R i e = if e ∈ G.incidence_set i then 1 else 0 := rfl
+
+/-! ### Relation between inc_matrix elements and incidence_set property -/
+
+lemma inc_matrix_coeff_eq_zero_iff {i : V} {e : sym2 V} [char_zero R] :
+  G.inc_matrix R i e = 0 ↔ e ∉ G.incidence_set i :=
+by simp only [inc_matrix, ite_eq_right_iff, ←decidable.not_imp_not,
+              forall_true_left, not_false_iff, one_ne_zero]
+
+lemma inc_matrix_coeff_eq_one_iff {i : V} {e : sym2 V} [char_zero R] :
+  G.inc_matrix R i e = 1 ↔ e ∈ G.incidence_set i :=
+by simp only [inc_matrix, ite_eq_left_iff, ←decidable.not_imp_not,
+              set.not_not_mem, forall_true_left, not_false_iff, zero_ne_one]
+
+/-! ### One - zero properties -/
+
+lemma inc_matrix_ne_zero {i : V} {e : sym2 V} [char_zero R] :
+  ¬ G.inc_matrix R i e = 0 ↔ G.inc_matrix R i e = 1 :=
+by simp only [inc_matrix_coeff_eq_zero_iff, inc_matrix_coeff_eq_one_iff, set.not_not_mem]
+
+lemma inc_matrix_ne_one {i : V} {e : sym2 V} [char_zero R] :
+  ¬ G.inc_matrix R i e = 1 ↔ G.inc_matrix R i e = 0 :=
+by simp only [inc_matrix_coeff_eq_zero_iff, inc_matrix_coeff_eq_one_iff]
+
+theorem adj_sum_of_mul_inc_one {i j : V} (H_adj : G.adj i j) :
+  ∑ e, G.inc_matrix R i e * G.inc_matrix R j e = (1 : R) :=
+begin
```

Comment:
Here's the proof without `adj_iff_exists_edge_coe`:
```
by simp only [inc_matrix_apply, ite_mul_ite_zero_right, sum_boole,
    G.mem_incidence_sets_iff_eq H_adj, mul_one, subtype.ext_iff.symm, filter_eq',
    filter_congr_decidable, if_true, mem_univ, nat.cast_one, card_singleton]
```
Definitely golfable if you feel like it.